Minimal Migration Procedure (WIP)

XedinUnknown opened the pull request:
Currently:

- Adds a `mlp2to3 relationships` command to WP CLI.
- Able to migrate a number of MLP2 link records to MLP3 relationships, one at a time.

XedinUnknown committed:
Added DI and i18n dependencies

XedinUnknown committed:
Added main files

These files facilitate the entry point to the application,
as well as a strutured approach to handlers and DI.

XedinUnknown committed:
Added missing `throws` doc

XedinUnknown committed:
Added `DatabaseWpdbTrait`

This abstracts database operations, allowing code to be
de-coupled from WPDB.

XedinUnknown committed:
Added dev dependency on WP CLI

Dinamiko committed:
Introduce acceptance test suite

XedinUnknown committed:
Added `IntegrationHandler`

This allows MLP3 to be activated while MLP2 is installed.
Requires the following changes:
https://bitbucket.org/inpsyde/multilingualpress/pull-requests/128/added-filter-to-bypass-legacy-check/commits

XedinUnknown committed:
Fixed bug in `_select()`

Used to cause error when using `prepare()` on queries with no
variables.

XedinUnknown committed:
Fixed some code standards

XedinUnknown committed:
Added `AddCliCommandCapableWpTrait`

XedinUnknown committed:
Added `MigrateCliCommandHandler`

XedinUnknown committed:
Added `wpdb` service

This eliminates the need to use the `wpdb` global

XedinUnknown committed:
Added `ContentRelationshipMigrator`

XedinUnknown committed:
Added `mlp2to3 relationships` WP CLI command

XedinUnknown committed:
Added missing `WpTriggerCapableTrait`

XedinUnknown committed:
Added files to be checked by PHPCS

Dinamiko commented:
Hi @XedinUnknown I did a first test with just default posts and one term connected, post relationships were created correctly, however no term relationship was created in MLP3:

![01](https://user-images.githubusercontent.com/456223/54817370-0b028d00-4c97-11e9-95a2-9aa489a1cc71.png)

I´ve run `wp mlp2to3 relationships`, should I pass any other parameter here related to terms? or is it a bug that the term relationship was not created? thanks.

XedinUnknown commented:
@Dinamiko, it's a bug related to the fact that the current implementation identifies content by the (`site_id` + `content_id`) pair, without taking into account the type of relationship.

Thank you for reporting this. The next iteration will take relationship type into account.

XedinUnknown committed:
Added `CliUtilsTrait`

XedinUnknown committed:
Added progress and error handling to `relationships` command

Importing will show a progress bar, and display a success
message with the amount of relationships migrated.

On failure, a red error message will be displayed.
In debug mode, a stack trace will be added to it.

XedinUnknown committed:
Relationship migration now includes type in link identification

This means that a single relationship is identified by
content ID, site ID, and the relationship type code, all together.

This should fix the problem where relationships for non-post
links were not getting created, because the site ID + content ID
was already migrated for post links.

XedinUnknown commented:
So, just to summarize, the latest changes include the relationship type in the identity of a relationship, so migrations for different types should now work. Also, I added a progress bar and some command output.

Looking forward to more feedback.

Dinamiko commented:
I did the same manual test again and now migration works for both `post` and `term` types.

XedinUnknown committed:
Added dependency `brain/monkey`

XedinUnknown committed:
Now depending on PHPUnit

XedinUnknown committed:
Moved reusable handler classes to their own directory

XedinUnknown committed:
Now using PHPUnit version 6

XedinUnknown committed:
Added PHPUnit config and bootstrap

The config will run all tests in the `func` and `unit` suites,
which have their respective directories under `test`.

XedinUnknown committed:
Now ignoring the test logs

XedinUnknown committed:
Now depending on Andrew

This allows tests to access protected members

XedinUnknown committed:
Added tests for `HandlerTrait`

XedinUnknown committed:
Added unit test for `RunHandlersCapableTrait`

XedinUnknown committed:
Moved reusable CLI classes to their own namespace

XedinUnknown committed:
Moved reusable config classes to their own namespace

XedinUnknown committed:
Moved DB related classes to their own namespace

XedinUnknown committed:
Moved event-related classes to their own namespace

XedinUnknown committed:
Moved re-usable migration classes to their own namespace

XedinUnknown committed:
Added unit tests for `WpHookingTrait`

XedinUnknown committed:
Added test for `WpTriggerCapableTrait`

widoz commented on main.php:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @return null|HandlerInterface
+ */
+function handler() {
```
Question: Why not use an iife to not expose constants and functions outside?

XedinUnknown replied:
Because you cannot invoke an IIFE later to retrieve the value cached in the `static` variable. This is the one thing that is coupled to the constants. Which is not that bad, given that both the function and the constants are plugin-specific and in the same file.

widoz replied:
But isn't supposed the handler called once?

widoz commented on main.php:
```diff
@@ -0,0 +1,41 @@
+function handler() {
+	static $instance = null;
+
+	if ( is_null( $instance ) ) {
```
Question: Advantages of `is_null` over `$instance === null`?
Question: Stupid thing but would not make more sense to have `if($instance !== null) { return $instance}` ?

XedinUnknown replied:
No advantages. Just the way I am used to. I guess, `is_null()` feels more like comparing the type, while  `$instance === null` feels more like comparing the value. I'm open to different ways, but don't see much use in choosing a specific one.

widoz commented on src/Migration/ContentRelationshipMigrator.php:
```diff
@@ -0,0 +1,262 @@
+        $typeCode
+    ): int {
+
+        if (!($groupId = $this->_retrieveGroupId($sourceBlogId, $sourceContentId, $typeCode))) {
```
Usually I do not like obfuscate assignment, moving it outside make it more visible that help during debug too.

XedinUnknown replied:
Such a simple expression as `$a = b()` does not seem like "obfuscation" to me. But if you prefer, I can expand this into two expressions.

widoz commented on src/MigrateRelationshipsCliCommand.php:
```diff
@@ -0,0 +1,122 @@
+                $this->migrator->migrate($relationship);
+                $progress->tick();
+            } catch (Exception $e) {
+                var_dump($e->getMessage());
```
Debug code on the run?

widoz commented on src/Migration/ContentRelationshipMigrator.php:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * {@inheritdoc}
+     */
+    protected function _getTranslator()
```
Sounds like this isn't necessary since the trait already implement it.

widoz commented on src/MigrateCliCommandHandler.php:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @param ContainerInterface $config The configuration of this handler.
+     */
+    public function __construct(ContainerInterface $config)
```
Why injecting the container here instead of simply what's needed by the class? Now the class has access at practically everything is within the container.

XedinUnknown replied:
Handler logic is basically controller level. It typically needs many things, and is prone to churn. To avoid both, I decided that this very application-specific layer can implement the Service Locator anti-pattern, and that's OK on this level. It also gives handlers the ability to lazy load the dependencies.

widoz commented on src/Cli/AddCliCommandCapableWpTrait.php:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return string The translated string.
+     */
+    abstract protected function __($string, $args = [], $context = null);
```
it's missing the return type hint. Found the same on other parts may be you want to run code inspection to catch them.

XedinUnknown replied:
That's true. However, if I add it, then the signature will conflict with that declared in the trait, which is meant to satisfy the contract.

widoz commented on src/Db/DatabaseWpdbTrait.php:
```diff
@@ -0,0 +1,102 @@
+        $insertId = $db->insert_id;
+
+        // Error while inserting
+        if (empty($result) && !empty($error)) {
```
Sounds like this isn't defined anywhere, could cause problems?

widoz commented on src/Migration/ContentRelationshipMigrator.php:
```diff
@@ -0,0 +1,262 @@
+    use StringTranslatorAwareTrait;
+
+    protected $db;
+    protected $translator;
```
Question: I see this defined within the Trait, I usually avoid to define properties within the traits because them hide information. I'm curious to know which is the best approach here.

XedinUnknown replied:
True, in many cases defining trait properties would hide information. Unless that trait is specifically responsible for read/write access (i.e. "awareness") of something, and implements it via a property. In that case, it's OK, and in fact there's no other way for that kind of trait to do this without beginning to assume things.

widoz reviewed (commented):
First of all, I don't know how the project will evolve so this review is about general impression I have about the implementation I can read.

At the end there is not too much to review, most of the code actually are dependencies that I cannot/want to review because out of the scope of this PR I guess. 

At first glance looks good to me, I think a bit over engineered for a plugin like this one and the approach used "non Inpsyde code" could be a problem in terms of new developer access, not so difficult but surely something to think about in future.

I'm not sure about the massive use of the Traits you do, sounds like more code to maintain that can be avoided, may something is done to make more simple unit tests or to wrap third party code that's in some circumstances could be useful but I don't get the point of having them for static methods or for internal getters. Could be that we are sacrificing implementation to be able to make tests more easily?

For some other things like extracting information from configuration I would instead encapsulate functionality within a class.

For example, regarding the hooks functions, why using a trait? Would make more sense to create a Service you can inject therefore you can use a different one? Wouldn't make the code more reusable? With a trait you are forced to use always the same. 

Even thought I don't get why the code of this plugin should be maximize to be reusable. But probably it's because you already have your own solution to solve specific problems. 

I didn't made a running test. I see @Dinamiko already did that and that's ok.

XedinUnknown commented:
> the approach used "non Inpsyde code" could be a problem in terms of new developer access, not so difficult but surely something to think about in future.

If you want to re-use open-source solutions, there will always be non-Inpsyde code. That's what re-use is about: using what other people wrote before us. IMO that's totally normal.

Currently, there are mainly 2 non-Inpsyde solutions being used:

1. DI container implementation, together with a memoizer for service caching. Very simple to set up: give it a map of service definitions, and retrieve services using PSR-11. Business as usual.
2. I18n wrapper. This centralizes things like the text domain, and provide an injectable standards-compliant (both in terms of interop, and `gettext`) solution for translations. This is slightly different to what most WP developers are used to, but we are not like most WP developers, and this isn't rocket science either - especially together with the fact that all translation in classes is further abstracted to the `__()` method. So, to the people reading class code, consumption of i18n functionality will look almost the same as before. Anyhow, all methods are documented. For me, pressing `Ctrl+Q` or `Ctrl+P` gives me all the info I need to know.

XedinUnknown commented:
> I'm not sure about the massive use of the Traits you do, sounds like more code to maintain that can be avoided, may something is done to make more simple unit tests or to wrap third party code that's in some circumstances could be useful but I don't get the point of having them for static methods or for internal getters. Could be that we are sacrificing implementation to be able to make tests more easily?

> For example, regarding the hooks functions, why using a trait? Would make more sense to create a Service you can inject therefore you can use a different one? Wouldn't make the code more reusable? With a trait you are forced to use always the same.

Although we had already talked about this before, I had a feeling you would ask this question :)

I agree that an injectable solution is superior to _just_ using traits like it is done here. However, the goal was to build an MVP as fast as possible. If we want to have real flexibility, we should use a standards-compliant event dispatching mechanism, which can be injected into the consumers, like you suggest. However, this requires research and much more consideration and work. For example, right now the only real interop standard available is the PSR-14 that was just published a few days ago. Obviously, right now there is no standards-compliant solution that would expose the WordPress hooks mechanism via an interop interface. The closest thing that exists is my [`dhii/wp-events`][] implementation. However, it was created a long time ago, when PSR-14 was not even in draft mode, and is not compliant with the current PSR-14 spec. Furthermore, using something like this would add yet another dependency to the project.

In that light, using a trait that declares e.g. `_addAction()` is a very simple and fast way to abstract the WP hook system away from consumers. This would allow us to later hook in `dhii/wp-events`, or a PSR-14-compliant implementation, or whatever else we want. Furthermore, the current traits are made in such a way to make them forward-compatible with the PSR-14 spec.

With regard to "always using the same" with traits. You are free to change the trait, and the abstraction would change. At some point, you are free to use a trait which uses an injected PSR-14 dispatcher, for example. Here, traits are not used _instead_ of injection, but are simply a way to further abstract - even from the interface! - and to defer architectural decisions to a later time, a time after MVP. When we decide to use e.g. an injected PSR-14 dispatcher, consuming code will not have to change thanks to this layer of abstraction.

I hope this explains the decision behind certain traits.


[`dhii/wp-events`]: https://github.com/Dhii/wp-events

XedinUnknown commented:
> At first glance looks good to me, I think a bit over engineered for a plugin like this one

This is something that I cannot really see, to be honest. The plugin consists of a bootstrap that configures the DI container, and mostly application-specific implementations. There are handlers that add the CLI command, initiate migration, provide integration, etc. There is a very very thin abstraction layer on top of WP-CLI and WP, etc. And then, there's the super-specific migration implementation. I don't see where the over-engineering is.

But if you could point me to some examples which you think are more complicated than necessary, I would be happy to explain, or consider changing.

XedinUnknown commented:
> For some other things like extracting information from configuration I would instead encapsulate functionality within a class.

I'm not sure what you mean. What functionality related to configuration is not encapsulated in a class? The retrieval of configuration is encapsulated in the DI container implementation.

XedinUnknown commented:
> Even thought I don't get why the code of this plugin should be maximize to be reusable. But probably it's because you already have your own solution to solve specific problems.

While it's true that I have my ways of solving problems, the re-usability here is but a desirable side-effect of Separation of Concerns, and other SOLID principles. I can never understand this: why write non-reusable code, when you can spend the same time on writing re-usable, or at least portable code?

widoz commented:
> > For some other things like extracting information from configuration I would instead encapsulate functionality within a class.
> 
> I'm not sure what you mean. What functionality related to configuration is not encapsulated in a class? The retrieval of configuration is encapsulated in the DI container implementation.

Refered to the use of `_getConfig` within the trait, but I guess you already answered in a previous comment.

Dinamiko committed:
Introduce content relationship acceptance test

Dinamiko committed:
Add acceptance test three sites relationship

XedinUnknown committed:
Re-factored for better semantic and architectural correctness

Dinamiko committed:
Move acceptance to test directory

Dinamiko committed:
Add multiple post and term relationship test (WIP)

Dinamiko committed:
Finish multiple posts and terms relationship test

XedinUnknown committed:
Re-factored running multiple handlers, and added `CompositeHandler`

XedinUnknown committed:
Re-factored running multiple handlers, and added `CompositeHandler`

XedinUnknown committed:
Added progress services

XedinUnknown committed:
Re-factored command and handlers

There now are layers of handler queues. The migration CLI command
creates a composite Migration handler, which will run Migration Module
handlers - one for every module. The Relationships module handler uses
the Relationships migrator to process a batch of Relationship records.

Each layer re-uses a Progress specific to that layer, which visually
displays the progress of that layer.

XedinUnknown committed:
Updated dependencies

XedinUnknown committed:
Added layered architecture

Migration handler now runs a handler per migration module.
Each layer has a progress bar of its own.

XedinUnknown committed:
Added migration for redirects

XedinUnknown committed:
Fixed service name

XedinUnknown committed:
Added missing handler for redirect migration

XedinUnknown committed:
Added missing redirect migrator

XedinUnknown committed:
Added limit support to redirect migration handler

Also, fixed return type doc.

XedinUnknown committed:
Fixed var name bug

XedinUnknown committed:
Fixed bug with migration args recognition

Now the migration modules should be recognized correctly.

XedinUnknown committed:
Corrected bug in string placeholder interpolation

XedinUnknown committed:
Added module status migration

Dinamiko committed:
Fix issue with hoa library (used by codeception)

Dinamiko committed:
Add acceptance test for modules

Dinamiko committed:
Add acceptance test for site redirection option

XedinUnknown committed:
Added language redirect migration

Dinamiko committed:
Add acceptance test for user language redirect

XedinUnknown committed:
Removed debug code

Dinamiko committed:
Add acceptance test for custom post types migration

XedinUnknown committed:
Added documentation to methods, and return type declaration to some.

XedinUnknown committed:
Abstracted setting network option

XedinUnknown committed:
Corrected return type doc

XedinUnknown committed:
Removed unused imports

XedinUnknown committed:
Added translatable post types migration

XedinUnknown committed:
Added usage doc to Readme

XedinUnknown committed:
Added documentation data to migration command

XedinUnknown committed:
Corrected and improved command doc

XedinUnknown committed:
Improved success message

Dinamiko committed:
Add acceptance test for site languages migration

Dinamiko committed:
Merge branch 'task/mvp' into tests/acceptance

Dinamiko committed:
Update acceptance tests

Dinamiko committed:
Add acceptance test for alternative language title

Dinamiko committed:
Add acceptance test for language repository migration

XedinUnknown committed:
Improved method names, making them more semantic

MLP-644

XedinUnknown committed:
Abstracted module name transformation

MLP-644

XedinUnknown committed:
It is now possible to directly map MLP2 module names to MLP3

Everything that is not in the map will follow standard transformation
rules.

MLP-644

XedinUnknown committed:
Added module name mapping for `redirect` module

MLP-644

XedinUnknown committed:
Removed syntax highlighting from usage doc example

Dinamiko committed:
Merge branch 'task/mvp' into tests/acceptance

Dinamiko committed:
Merge task/mvp into tests/acceptance

XedinUnknown committed:
Corrected license information

XedinUnknown committed:
Merge branch 'task/mvp' of https://github.com/inpsyde/multilingualpress-2-to-3-migration into task/mvp

XedinUnknown committed:
`0.1-alpha1` now points to `task/mvp` branch

XedinUnknown committed:
Switched license to GPLv2+

XedinUnknown committed:
Fixed branch alias

XedinUnknown committed:
Added fields and PK for languages table

XedinUnknown committed:
`DatabaseWpdbTrait` can now create tables

XedinUnknown committed:
Added `CreateTableHandler`

XedinUnknown committed:
Added `languages` module handler

It consists of multiple steps.

XedinUnknown committed:
Added table creation to `languages` steps

XedinUnknown committed:
Now depending on stringable intarface

XedinUnknown committed:
Now requiring `ext-json`

XedinUnknown committed:
Now outputting whole exception for debugging

Outputting only the trace used to prevent inner exceptions from
being displayed

XedinUnknown committed:
Now throwing if cannot prepare query

XedinUnknown committed:
Select query error messages now include query string

XedinUnknown committed:
Corrected param doc

XedinUnknown committed:
`CreateTableHandler` now uses table name transformation

This results in the table name being prefixed

XedinUnknown committed:
Added `plugins_dir` service

It calculates the directory where all plugins are stored, based on the
current plugin's basepath

XedinUnknown committed:
Added `is_debug` config option

XedinUnknown committed:
Added `FileContents` class and service

XedinUnknown committed:
Added `Json` and factory servie for it

XedinUnknown committed:
Added `Index` and service factory

XedinUnknown committed:
Added basic configuration reflecting MLP3 paths

XedinUnknown committed:
Added service to retrieve MLP3 langs JSON string

XedinUnknown committed:
Now retrieving MLP3 languages JSON object

XedinUnknown committed:
Now indexing the MLP3 embedded Locales and Languages separately

XedinUnknown committed:
Added methods for quoting identifiers and compiling fields strings

XedinUnknown committed:
Added `_updateBlogOption()`

This abstracts the write operation, also fixing a bug related to
undocumented `false` return in case of updating to same value.
An exception will now be thrown only if the operation actually fails.

See https://developer.wordpress.org/reference/functions/update_blog_option/#comment-3166

XedinUnknown committed:
Source fields of content relationships now aliased to a more normal form

XedinUnknown committed:
Changed indexing key for embedded languages and locales

Languages are now indexed by their ISO 639-3 code, and locales - by BCP47.

XedinUnknown committed:
Fixed but that didn't allow items to be excluded from index.

XedinUnknown committed:
Fixed bug in index

Caused the whole list to be added on match, instead of just the matched item.

XedinUnknown committed:
Corrected throws doc for `_insert()`

Now documents a more specific `Exception` too.

XedinUnknown committed:
Languages and Locales will no longer be added to index if they are missing
their ID field.

XedinUnknown committed:
Added the actual migration step to languages migration

XedinUnknown committed:
Removed `iso_639_3` col from language source selection

It doesn't exist in MLP3.

XedinUnknown committed:
Added `_query()` method

Just runs a generic query. Not capable of retrieving results.

XedinUnknown committed:
Added `_dropTable()`

XedinUnknown committed:
Added `_renameTable()`

XedinUnknown committed:
Added replacement of actual table to lang migration steps

XedinUnknown committed:
Added missing square brace

Used to cause a syntax error

XedinUnknown committed:
Added missing classes necessary for table replacement

XedinUnknown committed:
Can now pass arbitrary config from main file

All outside config is now passed that way, including some new keys.

XedinUnknown committed:
Now including `upgrade.php`

It contains functions useful for migrating.

XedinUnknown committed:
Added missing closing bracket

XedinUnknown committed:
Added smarter exception handling to main procedure

XedinUnknown committed:
Added more usage info to readme

XedinUnknown committed:
Added more installation detail, and requirements section

XedinUnknown committed:
Fixed bug that caused the source lang key to not be used for lang comparison

XedinUnknown committed:
Added source to dest key mapping for language comparison

This fixes a problem with the BCP47 not beig compared correctly.

XedinUnknown committed:
Added defaul values for destination language instanes

XedinUnknown committed:
Removed debug limit on nnumber of langs to migrate

XedinUnknown committed:
Added strict types declaration to `Json` file

XedinUnknown committed:
Improved code style

XedinUnknown committed:
`Json` now implements `StringableInterface`

XedinUnknown committed:
`Json` now allows integer keys

XedinUnknown committed:
`Json` can now access keys using property getters/setters

Thus, it mimicks a result of `json_decode()` more closely.

XedinUnknown committed:
Fixed bug that caused `Json` to incorrectly retrieve string value

The bug manifested if the `Json` instane was created with a
stringable object instead of string.

XedinUnknown committed:
Fixed but in string setter

XedinUnknown committed:
Now passing main site ID cofig

XedinUnknown committed:
Added `site_languages` migration module

XedinUnknown committed:
Updated readme with known limitations

Dinamiko committed:
Merge task/mvp into tests/acceptance

Dinamiko committed:
Merge branch 'task/mvp' into tests/acceptance

XedinUnknown committed:
Added missing migrator and handler

XedinUnknown committed:
Removed extra bracked, fixing syntax error

Dinamiko committed:
Working on languages acceptance test

XedinUnknown committed:
Fixed issue with wrong MLP3 site settings structure with `site_languages`

XedinUnknown committed:
Updated license code in main file

XedinUnknown committed:
Added license file

XedinUnknown committed:
Added WP containers for dealing with sites, options, and meta

XedinUnknown committed:
Fixed bug in site language migration

Used to put alternative site titles in site meta instead of options.
This fix uses the new WP containers to abstract options and meta
operations

Dinamiko committed:
Merge task/mvp into tests/acceptance

Dinamiko committed:
Merge task/mvp into tests/acceptance

XedinUnknown committed:
Removed accidental debug output

Dinamiko committed:
Merge branch 'task/mvp' into tests/acceptance

XedinUnknown committed:
Added a requirement for MLP2

XedinUnknown committed:
Corrected formatting of commit link

XedinUnknown committed:
Corrected requirement reference

XedinUnknown committed:
Corrected another requirement reference

XedinUnknown committed:
Merge branch 'task/mvp' into tests/acceptance

XedinUnknown committed:
Re-worked test setup

- Now everything will be in docker, with no additional host requirements.
- Moved configs to root, and fixed paths accordingly.
- Added `test` service for controlling PHP version etc.
- Added `codecept` and `chrome` services.
- Compatible with `docker-machine` by extracting basepath - see env file

XedinUnknown committed:
Site languages will now be migrated even if they already exist in target

XedinUnknown committed:
Now using private Packagist

XedinUnknown committed:
Now depending on MLP2 and MLP3 for dev

XedinUnknown committed:
Composer no longer complains about using root user

XedinUnknown committed:
Separated code semantically

XedinUnknown committed:
Now preventing deletion of shared tables

XedinUnknown committed:
Added "Development" section to readme

XedinUnknown committed:
Merge branch 'tests/acceptance' into task/mvp

XedinUnknown committed:
Fixed bug with wrong service name

XedinUnknown committed:
Switched environment to PHP 7.0.33

Now in line with minimal requirement

XedinUnknown committed:
Added autoload-dev rules

XedinUnknown committed:
Updated PHPS configs

XedinUnknown committed:
Added test helper trait

XedinUnknown committed:
Corrected name of test

XedinUnknown committed:
Added test for `IntegrationHandler`

XedinUnknown committed:
Corrected filter name

XedinUnknown committed:
Fixed shared table name

XedinUnknown committed:
Added another shared table name

XedinUnknown committed:
Added build instructions

XedinUnknown committed:
Updated PHPStorm configs

XedinUnknown committed:
Switched environment to PHP 7.0.33

This now corresponds to the minimal requirements

XedinUnknown committed:
Added "Building" section to readme

XedinUnknown committed:
Git archive will now ignore extra files

This results in those files not being present in the dist archive.

XedinUnknown committed:
Meta files will now be present in the plugin build

Most importantly, this causes the plugin to be shipped with a license.

XedinUnknown committed:
Added changelog

XedinUnknown commented:
No volunteers to review this more. However, @Dinamiko has run the acceptance tests in a local environment (not virtualized), and they pass. There has also been manual testing which shows that everything works as expected. Therefore, this is as good to go as it can be right now.
